core/factorization/par_ict.cpp

```diff
@@ -0,0 +1,307 @@
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/factorization/par_ict.hpp>
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/factorization/factorization_kernels.hpp"
+#include "core/factorization/par_ict_kernels.hpp"
+#include "core/factorization/par_ilu_kernels.hpp"
+#include "core/factorization/par_ilut_kernels.hpp"
+#include "core/matrix/coo_builder.hpp"
+#include "core/matrix/csr_builder.hpp"
+#include "core/matrix/csr_kernels.hpp"
+
+
+namespace gko {
+namespace factorization {
+namespace par_ict_factorization {
+
+
+GKO_REGISTER_OPERATION(threshold_select,
+                       par_ilut_factorization::threshold_select);
+GKO_REGISTER_OPERATION(threshold_filter,
+                       par_ilut_factorization::threshold_filter);
+GKO_REGISTER_OPERATION(threshold_filter_approx,
+                       par_ilut_factorization::threshold_filter_approx);
+GKO_REGISTER_OPERATION(add_candidates, par_ict_factorization::add_candidates);
+GKO_REGISTER_OPERATION(compute_factor, par_ict_factorization::compute_factor);
+
+GKO_REGISTER_OPERATION(initialize_row_ptrs_l,
+                       factorization::initialize_row_ptrs_l);
+GKO_REGISTER_OPERATION(initialize_l, factorization::initialize_l);
+
+GKO_REGISTER_OPERATION(csr_conj_transpose, csr::conj_transpose);
+GKO_REGISTER_OPERATION(convert_to_coo, csr::convert_to_coo);
+GKO_REGISTER_OPERATION(spgemm, csr::spgemm);
+
+
+}  // namespace par_ict_factorization
+
+
+using par_ict_factorization::make_add_candidates;
+using par_ict_factorization::make_compute_factor;
+using par_ict_factorization::make_convert_to_coo;
+using par_ict_factorization::make_csr_conj_transpose;
+using par_ict_factorization::make_initialize_l;
+using par_ict_factorization::make_initialize_row_ptrs_l;
+using par_ict_factorization::make_spgemm;
+using par_ict_factorization::make_threshold_filter;
+using par_ict_factorization::make_threshold_filter_approx;
+using par_ict_factorization::make_threshold_select;
+
+
+template <typename ValueType, typename IndexType>
+class ParIctState {
+    friend class ParIct<ValueType, IndexType>;
```

Comment:
Why make it a `class` and add `ParIct` as a `friend` ? Why not make it a `struct` instead ?